source/server/http/admin.cc

```diff
@@ -514,6 +514,45 @@ Http::Code AdminImpl::handlerCpuProfiler(absl::string_view url, Http::HeaderMap&
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerHeapProfiler(absl::string_view url, Http::HeaderMap&,
+                                          Buffer::Instance& response, AdminStream&) {
+  if (!Profiler::Heap::profilerEnabled()) {
+    response.add("The current build does not support heap profiler");
+    return Http::Code::InternalServerError;
+  }
+
+  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+  if (query_params.size() != 1 || query_params.begin()->first != "enable" ||
+      (query_params.begin()->second != "y" && query_params.begin()->second != "n")) {
+    response.add("?enable=<y|n>\n");
+    return Http::Code::BadRequest;
+  }
+
+  bool enable = query_params.begin()->second == "y";
+  if (enable) {
+    if (Profiler::Heap::isProfilerStarted()) {
+      response.add("Fail to start heap profiler: already started");
+      return Http::Code::BadRequest;
+    }
+    if (!Profiler::Heap::startProfiler(profile_path_)) {
+      response.add("Fail to start the heap profiler");
```

Comment:
Good catch! Will fix

Reply:
I plan to test whether profile_path_ is writable in the future. Could we leave the code here assuming `startProfiler` could return false some day?

Reply:
OK one thing you can do (again to avoid coverage issues and also to alert you if the situation arises in the future) is to put in `NOT_IMPLEMENTED_GCOVR_EXCL_LINE;` instead of the unreachable code. I think that will assert if it livens up some day.

Reply:
This remains open (add the NOT_IMPL... line), but otherwise lgtm.

Reply:
Is this block unreachable? If so, use `NOT_REACHED_GCOVR_EXCL_LINE` and remove the error handling code. Either of these macros cause Envoy to terminate so the modification of response and res will never execute.

However, it seems like starting the profiler might fail in the event that `profile_path_` is not valid and that doesn't seem like it should cause termination. To just exclude the lines from coverage tests, add comments like this:

```
// GCOVR_EXCL_START
response.add("Fail to start the heap profiler");
res = Http::Code::InternalServerError;
// GCOVR_EXCL_END
```
c.f. https://github.com/gcovr/gcovr/blob/master/doc/source/guide.rst#exclusion-markers

Reply:
@zuercher It's probably we are commenting on a obsolete commit. Not sure how github works.

Reply:
how or *if* :)